pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -393,10 +344,8 @@ struct RetryInfo {
 
 /// Data about the timeline to connect to, received from etcd.
 #[derive(Debug)]
-struct EtcdSkTimeline {
-    timeline: SkTimelineInfo,
-    /// Etcd generation, the bigger it is, the more up to date the timeline data is.
-    etcd_version: i64,
```

Comment:
Almost forgot to ask: before, this allowed us to get the latest timeline state out of the queue.
I did not notice anything similar in the broker code, so I wonder, how realistic is the situation when we start to "lag" on the receiving side (presumably, could happen since we're distracting on processing other signals from walreceiver)?
If that could happen, we could consider the node to be alive more than it actually is and get more latency.

Reply:
In theory it is possible, but in practice I don't think so. The cure for that could be comparing timestamps with sender, but this needs synchronized clocks and more code. However, reading the queue is cheap; so in such situation I think considering nodes too alive is a minor problem compared to general node overload.

Reply:
> comparing timestamps

Curious, why similar `broker_version: u64` could not be used? For our needs, that was very simple to use at least.
But sure if that's hard to do now, maybe not needed.

My general concern after all recent discussions is that we might get these corner cases in theory, but do not make them very visible?
For example, as I've learned [recently](https://github.com/neondatabase/neon/pull/2733#discussion_r1022586630) 

> If publisher can't send to subscriber, it drops the message. 

that could worsen the case of receiving some old data about the service from the broker subscription: our pageserver walreceiver heuristics would consider nodes alive more than needed, let's at least adjust those `trace!` messages to `warn!` (`error!`?) ones to be scrapped by default.

I think it would be also nice metrics on that and more logic around the `etcd_version: i64` and "skipping" somewhere, as it does not seem hard to do compared to the rest of the heavy-lifting 🙂 
But as long as we can notice such abnormalities, it's a good start already.

Reply:
> those trace! messages to warn!

I was somewhat barred by possibility of overflowing logs, but ok, let's do first version with that.

> why similar broker_version: u64 could not be used?

I don't see how it is useful here.

Reply:
I am thinking that walreceiver benefits from getting the freshmost updates to every SK node timeline data, while new model forces us to loop through all old messages and drops latest updates instead of the ones before: older data is obsolete for walreceiver.
If the node stops broadcasting broker updates, retaining its last messages also feels more natural: since messages are sent per second now, there could be big LSN increments between the skipped messages.

> possibility of overflowing logs

I think, we have log deduplication now, or at least the work was done on that front, and if I understand that right, ~ 85seconds (256 / 3) of some subscriber not reading the messages is also very odd and deserves logging.